File: src/main/java/com/paragon/client/systems/module/impl/render/Nametags.java
package com.paragon.client.systems.module.impl.render;

import com.paragon.Paragon;
import com.paragon.api.event.render.entity.RenderNametagEvent;
import com.paragon.api.util.entity.EntityUtil;
import com.paragon.api.util.player.EntityFakePlayer;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.ITextRenderer;
import com.paragon.asm.mixins.accessor.IRenderManager;
import com.paragon.api.module.Module;
import com.paragon.api.module.Category;
import com.paragon.client.systems.module.impl.client.ClientFont;
import com.paragon.api.setting.Setting;
import me.wolfsurge.cerauno.listener.Listener;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.ItemArmor;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemSword;
import net.minecraft.item.ItemTool;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.math.Vec3d;
import net.minecraft.util.text.TextFormatting;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import static org.lwjgl.opengl.GL11.*;

/**
 * @author Wolfsurge
 */
public class Nametags extends Module implements ITextRenderer {

    public static Nametags INSTANCE;

    // Render settings
    public static Setting<Boolean> health = new Setting<>("Health", true)
            .setDescription("Render the player's health");

    public static Setting<Boolean> ping = new Setting<>("Ping", true)
            .setDescription("Render the player's ping");

    public static Setting<Boolean> pops = new Setting<>("Pops", true)
            .setDescription("Render the player's totem pop count");

    public static Setting<Boolean> armour = new Setting<>("Armour", true)
            .setDescription("Render the player's armour");

    public static Setting<Boolean> armourDurability = new Setting<>("Durability", true)
            .setDescription("Render the player's armour durability")
            .setParentSetting(armour);

    public static Setting<Boolean> potions = new Setting<>("Potions", true)
            .setDescription("Shows the player's potion effects");

    // Scaling
    public static Setting<Float> scaleFactor = new Setting<>("Scale", 0.2f, 0.1f, 1f, 0.1f)
            .setDescription("The scale of the nametag");

    public static Setting<Boolean> distanceScale = new Setting<>("DistanceScale", true)
            .setDescription("Scale the nametag based on your distance from the player");

    public static Setting<Boolean> outline = new Setting<>("Outline", true)
            .setDescription("Render the nametag outline");

    public static Setting<Float> outlineWidth = new Setting<>("Width", 0.5f, 0.1f, 2f, 0.01f)
            .setDescription("The width of the outline")
            .setParentSetting(outline);

    public static Setting<Color> outlineColour = new Setting<>("Colour", new Color(185, 17, 255))
            .setDescription("The colour of the outline")
            .setParentSetting(outline);

    public Nametags() {
        super("Nametags", Category.RENDER, "Draws nametags above players");

        INSTANCE = this;
    }

    @Override
    public void onRender3D() {
        // Prevent null pointer exceptions
        if (nullCheck() || mc.player.connection == null || mc.player.ticksExisted < 20) {
            return;
        }

        // Iterate through loaded players
        for (EntityPlayer player : mc.world.playerEntities) {
            // Check the player isn't us
            if (player == mc.player) {
                continue;
            }

            // Get render x, y, and z
            double[] renderValues = {((IRenderManager) mc.getRenderManager()).getRenderX(), ((IRenderManager) mc.getRenderManager()).getRenderY(), ((IRenderManager) mc.getRenderManager()).getRenderZ()};
            // Get player interpolated position
            Vec3d renderVec = EntityUtil.getInterpolatedPosition(player);

            // Get scale
            double distance = mc.player.getDistance(renderVec.x, renderVec.y, renderVec.z);
            float scale = (scaleFactor.getValue() * 5) / 50f;

            if (distanceScale.getValue()) {
                scale = (float) (Math.max(scaleFactor.getValue() * 5, scaleFactor.getValue() * distance) / 50);
            }

            // Translate, rotate, and scale
            glPushMatrix();
            RenderHelper.enableStandardItemLighting();
            GlStateManager.disableLighting();
            glTranslated(renderVec.x - renderValues[0], renderVec.y + player.height + 0.1 + (player.isSneaking() ? 0.05 : 0.08) - renderValues[1], renderVec.z - renderValues[2]);
            glRotated(-mc.getRenderManager().playerViewY, 0, 1, 0);
            glRotated(mc.getRenderManager().playerViewX, (mc.gameSettings.thirdPersonView == 2) ? -1 : 1, 0, 0);
            glScaled(-scale, -scale, scale);

            // Disable depth so we can see the nametag through walls
            GlStateManager.disableDepth();

            StringBuilder stringBuilder = new StringBuilder(player.getName());

            if (health.getValue()) {
                stringBuilder.append(" ").append(EntityUtil.getTextColourFromEntityHealth(player)).append(Math.round(EntityUtil.getEntityHealth(player)));
            }

            if (ping.getValue() && mc.getConnection() != null) {
                mc.getConnection().getPlayerInfo(player.getUniqueID());
                stringBuilder.append(" ").append(getPingColour(mc.getConnection().getPlayerInfo(player.getUniqueID()).getResponseTime())).append(mc.getConnection().getPlayerInfo(player.getUniqueID()).getResponseTime());
            }

            if (pops.getValue()) {
                stringBuilder.append(" ").append(TextFormatting.GOLD).append("-").append((player instanceof EntityFakePlayer) ? 0 : Paragon.INSTANCE.getPopManager().getPops(player));
            }

            float potionWidth = 0;
            List<String> potionStrings = new ArrayList<>();

            player.getActivePotionEffects().stream().sorted(Comparator.comparing(PotionEffect::getDuration)).forEach(potion -> {
                        if (potions.getValue()) {
                            Potion potionType = Potion.getPotionFromResourceLocation(potion.getPotion().getRegistryName().toString());
                            if (potionType != null) {
                                potionStrings.add(potionType.getName().replace("potion.", "") + " " + potion.getDuration() + "s");
                            }
                        }
                    });

            potionStrings.sort(Comparator.comparingDouble(this::getStringWidth));

            // Get nametag width
            float width = getStringWidth(stringBuilder.toString()) + potionWidth + 4;

            // Center nametag
            glTranslated(-width / 2, -20, 0);

            // Draw background
            RenderUtil.drawRect(0, 0, width, getFontHeight() + (ClientFont.INSTANCE.isEnabled() ? 0 : 2), 0x90000000);

            // Draw border
            if (outline.getValue()) {
                RenderUtil.drawBorder(0, 0, width, getFontHeight() + (ClientFont.INSTANCE.isEnabled() ? 0 : 2), outlineWidth.getValue(), outlineColour.getValue().getRGB());
            }

            // Render string
            renderText(stringBuilder.toString(), 2, 2, -1);

            // Render armour
            if (armour.getValue()) {
                // Get the items we want to render
                ArrayList<ItemStack> stacks = new ArrayList<>();
                stacks.add(player.getHeldItemMainhand());
                Collections.reverse(player.inventory.armorInventory);
                stacks.addAll(player.inventory.armorInventory);
                Collections.reverse(player.inventory.armorInventory);
                stacks.add(player.getHeldItemOffhand());

                // Get armour count
                int count = 0;
                for (ItemStack stack : stacks) {
                    if (stack.isEmpty()) {
                        continue;
                    }

                    count++;
                }

                // Center armour
                int armourX = (int) ((width / 2) - ((count * 18) / 2));

                // Render armour
                for (ItemStack stack : stacks) {
                    // Check the stack isn't empty
                    if (stack.getItem() != Items.AIR) {
                        // Y value
                        int y = -20;

                        glPushMatrix();
                        glDepthMask(true);
                        GlStateManager.clear(256);
                        GlStateManager.disableDepth();
                        GlStateManager.enableDepth();
                        RenderHelper.enableStandardItemLighting();
                        mc.getRenderItem().zLevel = -100.0F;
                        GlStateManager.scale(1, 1, 0.01f);

                        // Render the armour
                        mc.getRenderItem().renderItemAndEffectIntoGUI(stack, armourX, y);
                        mc.getRenderItem().renderItemOverlays(mc.fontRenderer, stack, armourX, y);

                        mc.getRenderItem().zLevel = 0.0F;
                        GlStateManager.scale(1, 1, 1);
                        RenderHelper.disableStandardItemLighting();
                        GlStateManager.enableAlpha();
                        GlStateManager.disableBlend();
                        GlStateManager.disableLighting();
                        GlStateManager.scale(0.5D, 0.5D, 0.5D);
                        GlStateManager.disableDepth();

                        List<String> itemInfo = new ArrayList<>();

                        // Render the armour's durability
                        if (armourDurability.getValue() && stack.getItem() instanceof ItemArmor || stack.getItem() instanceof ItemSword || stack.getItem() instanceof ItemTool) {
                            float damage = 1 - (float) stack.getItemDamage() / (float) stack.getMaxDamage();
                            itemInfo.add((int) (damage * 100) + "%");
                        }

                        // Render enchants
                        NBTTagList enchants = stack.getEnchantmentTagList();
                        for (int i = 0; i < enchants.tagCount(); i++) {
                            // Get enchant ID and level
                            int id = enchants.getCompoundTagAt(i).getInteger("id");
                            int level = enchants.getCompoundTagAt(i).getInteger("lvl");

                            // Get the enchantment
                            Enchantment enchantment = Enchantment.getEnchantmentByID(id);

                            // Make sure the enchantment is valid
                            if (enchantment != null) {
                                // Don't render if it's a curse
                                if (enchantment.isCurse()) {
                                    continue;
                                }

                                // Get enchantment name
                                String enchantmentName = enchantment.getTranslatedName(level).substring(0, 4) + (level == 1 ? "" : level);

                                // Render the enchantment's name and level
                                itemInfo.add(enchantmentName);
                            }
                        }

                        float yOffset = -(itemInfo.size() * getFontHeight()) + 15;
                        for (String info : itemInfo) {
                            int colour = -1;

                            if (info.equals(itemInfo.get(0))) {
                                float green = ((float) stack.getMaxDamage() - (float) stack.getItemDamage()) / (float) stack.getMaxDamage();
                                float red = 1 - green;

                                colour = new Color(red, green, 0, 1).getRGB();
                            }

                            renderText(info, armourX * 2 + 4, y + yOffset, colour);
                            yOffset += getFontHeight();
                        }

                        // Re enable depth
                        GlStateManager.enableDepth();
                        GlStateManager.scale(2.0F, 2.0F, 2.0F);
                        glPopMatrix();

                        // Increase X value
                        armourX += 18;
                    }
                }
            }

            float y = 0;

            // GL11.glScalef(0.75f, 0.75f, 0.75f);
            // float scaleFactor = 1 / 0.75f;

            for (String effectString : potionStrings) {
                renderText(effectString, (getStringWidth(stringBuilder.toString()) + 6), y, -1);

                y += getFontHeight();
            }

            // End render
            GlStateManager.enableDepth();
            GlStateManager.disableBlend();
            glPopMatrix();
        }
    }

    public TextFormatting getPingColour(int ping) {
        if (ping < 0) {
            return TextFormatting.RED;
        } else if (ping < 50) {
            return TextFormatting.GREEN;
        } else if (ping < 150) {
            return TextFormatting.YELLOW;
        } else {
            return TextFormatting.RED;
        }
    }

    @Listener
    public void onRenderNametag(RenderNametagEvent event) {
        if (event.getEntity() instanceof EntityPlayer) {
            event.cancel();
        }
    }
}

File: src/main/kotlin/com/paragon/client/systems/module/impl/render/Nametags.kt
@file:Suppress("ReplaceNotNullAssertionWithElvisReturn")

package com.paragon.client.systems.module.impl.render

import com.paragon.Paragon
import com.paragon.api.event.render.entity.RenderNametagEvent
import com.paragon.api.module.Category
import com.paragon.api.module.Module
import com.paragon.api.setting.Setting
import com.paragon.api.util.anyNull
import com.paragon.api.util.entity.EntityUtil
import com.paragon.api.util.player.EntityFakePlayer
import com.paragon.api.util.render.RenderUtil.drawBorder
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.api.util.render.font.FontUtil.drawStringWithShadow
import com.paragon.api.util.render.font.FontUtil.getHeight
import com.paragon.api.util.render.font.FontUtil.getStringWidth
import com.paragon.mixins.accessor.IRenderManager
import com.paragon.bus.listener.Listener
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.enchantment.Enchantment
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.item.ItemArmor
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemSword
import net.minecraft.item.ItemTool
import net.minecraft.util.text.TextFormatting
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.util.*
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * @author Surge
 */
object Nametags : Module("Nametags", Category.RENDER, "Draws nametags above players") {

    // Render settings
    private val health = Setting(
        "Health",
        true
    ) describedBy "Render the player's health"

    private val ping = Setting(
        "Ping",
        true
    ) describedBy "Render the player's ping"

    private val pops = Setting(
        "Pops",
        true
    ) describedBy "Render the player's totem pop count"

    private val armour = Setting(
        "Armour",
        true
    ) describedBy "Render the player's armour"

    private val armourDurability = Setting(
        "Durability",
        true
    ) describedBy "Render the player's armour durability" subOf armour

    // Scaling
    private val scaleFactor = Setting(
        "Scale",
        0.2f,
        0.1f,
        1f,
        0.1f
    ) describedBy "The scale of the nametag"

    private val distanceScale = Setting(
        "DistanceScale",
        true
    ) describedBy "Scale the nametag based on your distance from the player"

    private val outline = Setting(
        "Outline",
        true
    ) describedBy "Render the nametag outline"

    private val outlineWidth = Setting(
        "Width",
        0.5f,
        0.1f,
        2f,
        0.01f
    ) describedBy "The width of the outline" subOf outline

    private val outlineColour = Setting(
        "Colour",
        Color(185, 17, 255)
    ) describedBy "The colour of the outline" subOf outline

    override fun onRender3D() {
        // Prevent null pointer exceptions
        if (minecraft.anyNull || minecraft.player.connection == null || minecraft.player.ticksExisted < 20) {
            return
        }

        // Iterate through loaded players
        for (player in minecraft.world.playerEntities) {
            // Check the player isn't us
            if (player === minecraft.player) {
                continue
            }

            // Get render x, y, and z
            val renderValues = doubleArrayOf(
                (minecraft.renderManager as IRenderManager).renderX,
                (minecraft.renderManager as IRenderManager).renderY,
                (minecraft.renderManager as IRenderManager).renderZ
            )
            // Get player interpolated position
            val renderVec = EntityUtil.getInterpolatedPosition(player)

            // Get scale
            val distance = minecraft.player.getDistance(renderVec.x, renderVec.y, renderVec.z)
            var scale = scaleFactor.value * 5 / 50f
            if (distanceScale.value) {
                scale = (max((scaleFactor.value * 5).toDouble(), scaleFactor.value * distance) / 50).toFloat()
            }

            // Translate, rotate, and scale
            GL11.glPushMatrix()
            RenderHelper.enableStandardItemLighting()
            GlStateManager.disableLighting()
            GL11.glTranslated(
                renderVec.x - renderValues[0],
                renderVec.y + player.height + 0.1 + (if (player.isSneaking) 0.05 else 0.08) - renderValues[1],
                renderVec.z - renderValues[2]
            )
            GL11.glRotated(-minecraft.renderManager.playerViewY.toDouble(), 0.0, 1.0, 0.0)
            GL11.glRotated(
                minecraft.renderManager.playerViewX.toDouble(),
                (if (minecraft.gameSettings.thirdPersonView == 2) -1 else 1).toDouble(),
                0.0,
                0.0
            )
            GL11.glScaled(-scale.toDouble(), -scale.toDouble(), scale.toDouble())

            // Disable depth so we can see the nametag through walls
            GlStateManager.disableDepth()
            val stringBuilder = StringBuilder(player.name)
            if (health.value) {
                stringBuilder.append(" ").append(EntityUtil.getTextColourFromEntityHealth(player))
                    .append(EntityUtil.getEntityHealth(player).roundToInt())
            }
            if (ping.value && minecraft.connection != null) {
                minecraft.connection!!.getPlayerInfo(player.uniqueID)
                stringBuilder.append(" ")
                    .append(getPingColour(minecraft.connection!!.getPlayerInfo(player.uniqueID).responseTime)).append(
                        minecraft.connection!!.getPlayerInfo(player.uniqueID).responseTime
                    )
            }
            if (pops.value) {
                stringBuilder.append(" ").append(TextFormatting.GOLD).append("-")
                    .append(if (player is EntityFakePlayer) 0 else Paragon.INSTANCE.popManager.getPops(player))
            }

            // Get nametag width
            val width = getStringWidth(stringBuilder.toString())

            // Center nametag
            GL11.glTranslated((-width / 2).toDouble(), -20.0, 0.0)

            // Draw background
            drawRect(0f, 0f, width, getHeight() + if (isEnabled) 0 else 2, -0x70000000)

            // Draw border
            if (outline.value) {
                drawBorder(
                    0f,
                    0f,
                    width,
                    getHeight() + if (isEnabled) 0 else 2,
                    outlineWidth.value,
                    outlineColour.value.rgb
                )
            }

            // Render string
            drawStringWithShadow(stringBuilder.toString(), 2f, 2f, -1)

            // Render armour
            if (armour.value) {
                // Get the items we want to render
                val stacks = ArrayList<ItemStack>()
                stacks.add(player.heldItemMainhand)
                Collections.reverse(player.inventory.armorInventory)
                stacks.addAll(player.inventory.armorInventory)
                Collections.reverse(player.inventory.armorInventory)
                stacks.add(player.heldItemOffhand)

                // Get armour count
                var count = 0
                for (stack in stacks) {
                    if (stack.isEmpty) {
                        continue
                    }
                    count++
                }

                // Center armour
                var armourX = (width / 2 - count * 18 / 2).toInt()

                // Render armour
                for (stack in stacks) {
                    // Check the stack isn't empty
                    if (stack.item !== Items.AIR) {
                        // Y value
                        val y = -20
                        GL11.glPushMatrix()
                        GL11.glDepthMask(true)
                        GlStateManager.clear(256)
                        GlStateManager.disableDepth()
                        GlStateManager.enableDepth()
                        RenderHelper.enableStandardItemLighting()
                        minecraft.renderItem.zLevel = -100.0f
                        GlStateManager.scale(1f, 1f, 0.01f)

                        // Render the armour
                        minecraft.renderItem.renderItemAndEffectIntoGUI(stack, armourX, y)
                        minecraft.renderItem.renderItemOverlays(minecraft.fontRenderer, stack, armourX, y)
                        minecraft.renderItem.zLevel = 0.0f
                        GlStateManager.scale(1f, 1f, 1f)
                        RenderHelper.disableStandardItemLighting()
                        GlStateManager.enableAlpha()
                        GlStateManager.disableBlend()
                        GlStateManager.disableLighting()
                        GlStateManager.scale(0.5, 0.5, 0.5)
                        GlStateManager.disableDepth()
                        val itemInfo: MutableList<String> = ArrayList()

                        // Render the armour's durability
                        if (armourDurability.value && stack.item is ItemArmor || stack.item is ItemSword || stack.item is ItemTool) {
                            val damage = 1 - stack.itemDamage.toFloat() / stack.maxDamage.toFloat()
                            itemInfo.add((damage * 100).toInt().toString() + "%")
                        }

                        // Render enchants
                        val enchants = stack.enchantmentTagList
                        for (i in 0 until enchants.tagCount()) {
                            // Get enchant ID and level
                            val id = enchants.getCompoundTagAt(i).getInteger("id")
                            val level = enchants.getCompoundTagAt(i).getInteger("lvl")

                            // Get the enchantment
                            val enchantment = Enchantment.getEnchantmentByID(id)

                            // Make sure the enchantment is valid
                            if (enchantment != null) {
                                // Don't render if it's a curse
                                if (enchantment.isCurse) {
                                    continue
                                }

                                // Get enchantment name
                                val enchantmentName =
                                    enchantment.getTranslatedName(level).substring(0, 4) + if (level == 1) "" else level

                                // Render the enchantment's name and level
                                itemInfo.add(enchantmentName)
                            }
                        }
                        var yOffset = -(itemInfo.size * getHeight()) + 15
                        for (info in itemInfo) {
                            var colour = -1
                            if (info == itemInfo[0]) {
                                val green =
                                    (stack.maxDamage.toFloat() - stack.itemDamage.toFloat()) / stack.maxDamage.toFloat()
                                val red = 1 - green
                                colour = Color(red, green, 0f, 1f).rgb
                            }
                            drawStringWithShadow(info, (armourX * 2 + 4).toFloat(), y + yOffset, colour)
                            yOffset += getHeight()
                        }

                        // Re enable depth
                        GlStateManager.enableDepth()
                        GlStateManager.scale(2.0f, 2.0f, 2.0f)
                        GL11.glPopMatrix()

                        // Increase X value
                        armourX += 18
                    }
                }
            }

            // End render
            GlStateManager.enableDepth()
            GlStateManager.disableBlend()
            GL11.glPopMatrix()
        }
    }

    private fun getPingColour(ping: Int): TextFormatting {
        return when {
            ping < 0 -> TextFormatting.RED
            ping < 50 -> TextFormatting.GREEN
            ping < 150 -> TextFormatting.YELLOW
            else -> TextFormatting.RED
        }
    }

    @Listener
    fun onRenderNametag(event: RenderNametagEvent) {
        if (event.entity is EntityPlayer) {
            event.cancel()
        }
    }

}